### PR TITLE
iso9660: rework IFile and IFileDirectory implementations

### PR DESCRIPTION
**Pull request by lrusak**
This changes the iso9660 methods to use libcdio++ and libiso9660++

This allows using the CURL hostname trick that is done for udf files. (I'm not sure why they were treated differently?)

These methods are only using for browsing into isos with an iso9660 and reading files from an iso with an iso9660 filesystem. These methods are not used for video playback (as that is done by libdvd).

These methods are mostly unused but can be used to browse music iso's or video iso files that don't contain proper dvd file structure.

depends will need to be built and uploaded for windows. All that is needed is to enable the `--enable-cxx` for libcdio to build libcdio++ and libiso9660++

I tested this using iso9660 iso's containing mp3 files that was made with `mkisofs`

This is part of a larger cleanup to get rid of using windows api globally.

### CODE REVIEW

**Comment by Rechi** [comment]
Debian (and Ubuntu) doesn't provide libcdio++ and `libiso9660++.

**Comment by lrusak** [comment]
> Debian (and Ubuntu) doesn't provide libcdio++ and `libiso9660++.

well that will have to get fixed then :smile: fedora packages them all together
```
$ rpm -qf /usr/lib64/libcdio++.so.1.0.0                                                                                                                                                                              
libcdio-2.0.0-4.fc30.x86_64
```

another one for the team ppa?

**Comment by wsnipex** [comment]
If I have to package libcdio, then I'd go for latest stable v2.1. In that case we should bump our version here as well

**Comment by wsnipex** [comment]
untested libcdio v2.1 with c++ bindings: https://launchpad.net/%7Ewsnipex/+archive/ubuntu/test/+packages?field.name_filter=cdio&field.status_filter=published&field.series_filter=

**Comment by lrusak** [comment]
jenkins build this please

**Comment by lrusak** [comment]
Waiting on windows. Who can build and upload the binaries?

**Review by Rechi — CHANGES_REQUESTED** [review]
As at least cdio version 0.94 is required with this PR, https://github.com/xbmc/xbmc/tree/master/tools/depends/target/libcdio has to be dropped.
Also docs have to be updated and tools/depends/target/libcdio-gplv3 needs to be buildable without depends (like crossguid, flatbuffers, libfmt, rapidjson, wayland-protocols and waylandpp).

File: xbmc/URL.cpp [hunk 1]
```diff
@@ -205,45 +205,42 @@ void CURL::Parse(const std::string& strURL1)
...
 
-      iPos = iAlphaSign + 1;
-      iSlash = strURL.find("/", iAlphaSign);
+    iPos = iAlphaSign + 1;
+    iSlash = strURL.find("/", iAlphaSign);
 
-      if(iSlash >= iEnd)
-        iSlash = std::string::npos;
-    }
+    if(iSlash >= iEnd)
```
Comment (Rechi) [inline xbmc/URL.cpp]: space messing between `if` and `(`

File: CMakeLists.txt [hunk 2]
```diff
@@ -51,6 +51,7 @@ option(ENABLE_PYTHON      "Enable python support?" ON)
 option(ENABLE_INTERNAL_FFMPEG "Enable internal ffmpeg?" OFF)
 if(UNIX)
   option(FFMPEG_PATH        "Path to external ffmpeg?" "")
+  option(ENABLE_INTERNAL_CDIO "Enable internal cdio?" ON)
```
Comment (Rechi) [inline CMakeLists.txt]: Why enable the internal version by default?

File: cmake/modules/FindCdiopp.cmake [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+                                  PATHS ${PC_CDIO_libiso9660_INCLUDEDIR})
+
+  find_library(ISO9660_LIBRARY NAMES libiso9660 iso9660
+                                 PATHS ${PC_CDIO_libiso9660_LIBDIR})
+
+  # end todo
+
+  set(CDIOPP_VERSION ${PC_CDIOPP_libcdio++_VERSION})
+
+  if(PC_CDIOPP_FOUND)
```
Comment (Rechi) [inline cmake/modules/FindCdiopp.cmake]: If pkg-config found `libcdio++`, you still do not know if `find_path` and `find_library` found something

**Review by Rechi — CHANGES_REQUESTED** [review]
Updating documentation is still missing.
`tools/depends/target/libcdio` has to to be buildable without depends (like crossguid, flatbuffers, libfmt, rapidjson, wayland-protocols and waylandpp).

File: CMakeLists.txt (hunk 2 above)
Comment (lrusak) [reply CMakeLists.txt]: because it's not optional

Comment (Rechi) [reply CMakeLists.txt]: That's not a reason for enabling the internal version by default.

Comment (lrusak) [reply CMakeLists.txt]: What's the point then?

File: cmake/modules/FindCdiopp.cmake [hunk 4]
```diff
@@ -99,9 +99,14 @@ else()
 
   # end todo
 
+  if(CDIOPP_INCLUDE_DIR AND CDIOPP_LIBRARY AND ISO9660PP_INCLUDE_DIR AND ISO9660PP_LIBRARY AND
+     CDIO_INCLUDE_DIR AND CDIO_LIBRARY AND ISO9660_INCLUDE_DIR AND ISO9660_LIBRARY)
+    set(CDIOPP_FOUND TRUE)
+  endif()
```
Comment (Rechi) [inline cmake/modules/FindCdiopp.cmake]: no hacks, use `find_package_handle_standard_args(...)`

File: tools/depends/target/libcdio/Makefile [hunk 5]
```diff
@@ -11,6 +11,22 @@ CONFIGURE=./configure --prefix=$(PREFIX)  --with-cd-drive=no --with-cd-info=no \
           --with-cdda-player=no --with-cd-read=no --with-iso-info=no --with-iso-read=no --disable-example-progs \
           --disable-cpp-progs --enable-cxx --disable-shared
 
+ifeq ($(CROSS_COMPILING), yes)
+  DEPS += ../../Makefile.include
+  CXXFLAGS += -std=c++11
```
Comment (Rechi) [inline tools/depends/target/libcdio/Makefile]: Why is this now needed for a complete depends build?

File: CMakeLists.txt (hunk 2 above)
Comment (Rechi) [reply CMakeLists.txt]: Disable it by default.

File: docs/README.Linux.md [hunk 6]
```diff
@@ -8,14 +8,14 @@ Several distribution **[specific build guides](README.md)** are available.
 ## Table of Contents
 1. **[Document conventions](#1-document-conventions)**
 2. **[Get the source code](#2-get-the-source-code)**
-3. **[Install the required packages](#3-install-the-required-packages)**
```
Comment (Rechi) [inline docs/README.Linux.md]: removing the spaces breaks the layout, see https://github.com/lrusak/xbmc/blob/iso9660-rework/docs/README.Linux.md

**Review by Rechi — CHANGES_REQUESTED** [review]
I think only updating  `docs/README.Linux.md` is not enough.

File: tools/depends/target/libcdio/Makefile (hunk 5 above)
Comment (lrusak) [reply tools/depends/target/libcdio/Makefile]: it's copy/paste from crossguid

File: cmake/modules/FindCdiopp.cmake [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+  find_path(ISO9660_INCLUDE_DIR NAMES cdio/iso9660.h
+                                  PATHS ${PC_CDIO_libiso9660_INCLUDEDIR})
+
+  find_library(ISO9660_LIBRARY NAMES libiso9660 iso9660
+                                 PATHS ${PC_CDIO_libiso9660_LIBDIR})
+
+  # end todo
+
+  set(CDIOPP_VERSION ${PC_CDIOPP_libcdio++_VERSION})
```
Comment (Rechi) [inline cmake/modules/FindCdiopp.cmake]: `PC_CDIOPP_VERSION` or `PC_CDIO_VERSION` instead of `PC_CDIOPP_libcdio++_VERSION`

**Comment by lrusak** [comment]
@notspiff can you take a look here if you are happy with the behaviour? My aim here now is to make the iso9660++ stuff optional as it depends on the gplv3 version of libcdio. With this being optional like I've made it we can split it off into a vfs binary-addon (I'd like to finish with the rest of the cdio stuff before I do that). Thoughts?

**Pull request merged**